packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,792 @@
+export interface Vis {
+  canvas: HTMLCanvasElement;
+  colors: string[];
+  analyser?: AnalyserNode;
+  oscStyle?: "dots" | "solid" | "lines";
+  bandwidth?: "wide" | "thin";
+  coloring?: "fire" | "line" | "normal";
+  peaks?: boolean;
+  saFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  saPeakFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa?: "analyzer" | "oscilloscope" | "none";
+  renderHeight?: number;
+  smallVis?: boolean;
+  pixelDensity?: number;
+  doubled?: boolean;
+  isMWOpen?: boolean;
+}
+import { FFT } from "./FFTNullsoft";
+
+/**
+ * Base class of Visualizer (animation frame renderer engine)
+ */
+export class VisPaintHandler {
```

Comment:
Could probably make this an abstract class https://www.tutorialsteacher.com/typescript/abstract-class